File: 프로그래머스/LV2/코딩테스트 연습/연습문제/롤케이크 자르기/롤케이크 자르기.py
def solution(topping):
    answer = 0
    ln,rn = 0,len(set(topping))
    left = set()
    right = dict()
    for t in topping:
        if t not in right: right[t] = 1
        else: right[t] += 1

    for i in range(len(topping)-1):
        if i != 0 and ln == rn:
            answer += 1

        if topping[i] not in left:
            ln += 1
        left.add(topping[i])

        right[topping[i]] -= 1
        if right[topping[i]] == 0:
            rn -= 1

    return answer

File: 프로그래머스/LV2/코딩테스트 연습/연습문제/롤케이크 자르기/롤케이크 자르기.cpp
#include <string>
#include <vector>
#include <unordered_set>
#include <unordered_map>
using namespace std;

int solution(vector<int> topping) {
    int answer = 0;
    int ln = 0, rn = 0;
    unordered_set<int> left;
    unordered_map<int, int> right;

    for(int t : topping) {
        if(right.find(t) == right.end()) {
            right[t] = 1;
            rn++;
        }
        else right[t]++;
    }

    for(int i = 0; i < topping.size()-1; i++) {
        if(i != 0 && ln == rn)answer++;

        if(left.find(topping[i]) == left.end())ln++;
        left.insert(topping[i]);

        right[topping[i]]--;
        if(right[topping[i]] == 0)rn--;
    }

    return answer;
}